### PR TITLE
Allow receipt recovery if the upload fails

### PR DESCRIPTION
**Pull request by Gonals**
### Fixed Issues

$ https://github.com/Expensify/App/issues/28884
PROPOSAL: 

### Tests
1. Open a chat in NewDot and click `+` -> `Request Money` -> `Scan` and upload a file that will fail in the backend. A CSV should do the trick.
2. Open the Money Request report chat. After a second or two, the request should fail and you should see this message:
Screenshot 2023-10-17 at 5 22 42 PM

3. Click on it and confirm you can download the file (Note that. refreshing or closing the app will make the file unavailable)

- [x] Verify that no errors appear in the JS console

### Offline tests
No changes

### QA Steps
Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot 2023-10-18 at 4 44 54 PM

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot 2023-10-18 at 4 42 20 PM

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Screenshot 2023-10-18 at 4 34 07 PM

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2023-10-17 at 5 22 52 PM

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2023-10-18 at 1 48 16 PM

</details>

### CODE REVIEW

**Comment by Gonals** [comment]
I'm having some trouble uploading receipts on mobile, but this is ready for testing!

**Comment by trjExpensify** [comment]
Ah, @Gonals.. I just realised that there's a small design addition when dismissing the error that we landed on with replace receipt flow where we're also wanting to do similar to not lose the receipt. It's the "are you sure?" modal confirmation on dismissing the error to be extra sure, illustrated [here](https://github.com/Expensify/App/issues/28343#issuecomment-1747716540). 

@tienifr is working on a PR [for that issue](https://github.com/Expensify/App/issues/28343#issuecomment-1738414257) in the replace receipt flow to not lose the receipt there on replacing it. Not sure if he needs to wait until this is merged to re-use parts of it, and then maybe aslo include adding that additional modal in this flow as well actually? 

CC: @hoangzinh

**Comment by Gonals** [comment]
> Ah, @Gonals.. I just realised that there's a small design addition when dismissing the error that we landed on with replace receipt flow where we're also wanting to do similar to not lose the receipt. It's the "are you sure?" modal confirmation on dismissing the error to be extra sure, illustrated [here](https://github.com/Expensify/App/issues/28343#issuecomment-1747716540).
> 
> @tienifr is working on a PR [for that issue](https://github.com/Expensify/App/issues/28343#issuecomment-1738414257) in the replace receipt flow to not lose the receipt there on replacing it. Not sure if he needs to wait until this is merged to re-use parts of it, and then maybe aslo include adding that additional modal in this flow as well actually?
> 
> CC: @hoangzinh

The don't need to wait on this. This flow is error-based, while theirs is simply adding the link in the modal, so they don't really intersect.

Similarly, we can't really add the modal here. I `guess` we could add it before dismissing the error, but I think that would be overkill and not really very useful, given than closing the app or refreshing the page will make them lose the receipt anyways (nothing we can do about that. It is stored in the temporary memory)

**Comment by Gonals** [comment]
> > Ah, @Gonals.. I just realised that there's a small design addition when dismissing the error that we landed on with replace receipt flow where we're also wanting to do similar to not lose the receipt. It's the "are you sure?" modal confirmation on dismissing the error to be extra sure, illustrated [here](https://github.com/Expensify/App/issues/28343#issuecomment-1747716540).
> > @tienifr is working on a PR [for that issue](https://github.com/Expensify/App/issues/28343#issuecomment-1738414257) in the replace receipt flow to not lose the receipt there on replacing it. Not sure if he needs to wait until this is merged to re-use parts of it, and then maybe aslo include adding that additional modal in this flow as well actually?
> > CC: @hoangzinh
> 
> The don't need to wait on this. This flow is error-based, while theirs is simply adding the link in the modal, so they don't really intersect.
> 
> Similarly, we can't really add the modal here. I `guess` we could add it before dismissing the error, but I think that would be overkill and not really very useful, given than closing the app or refreshing the page will make them lose the receipt anyways (nothing we can do about that. It is stored in the temporary memory)

Honestly, I don't think we even need the "download" link in the modal for receipt replacement. If they want it, they can close it and download it on the right hand corner. It is an extra click for a pretty uncommon flow 🤷

**Comment by trjExpensify** [comment]
We talked about it in thread and because dismissing the error message by clicking the `X` is irreversible to lose the receipt, it made sense for an "Are you sure?" prompt if you click the X to be sure before they take that action.

**Comment by trjExpensify** [comment]
>This flow is error-based, while theirs is simply adding the link in the modal, so they don't really intersect.

The other flow is as well right? Right now absolutely nothing happens when you replace a receipt with a dodgy file. So we need that issue to one show the error and two offer the ability to download the failed receipt replacement.

**Comment by Gonals** [comment]
> The other flow is as well right? Right now absolutely nothing happens when you replace a receipt with a dodgy file. So we need that issue to one show the error and two offer the ability to download the failed receipt replacement.

Ah, yes, I was thinking about the "Are you sure...?" modal

**Comment by trjExpensify** [comment]
Right, that issue is adding the full error handling flow incl. receipt download:

1. Adds the RBR error which can be dismissed
2. Stores the receipt locally to make it available to download
3. Adds the confirmation modal on dismiss of the error

**Comment by Gonals** [comment]
@0xmiroslav bump on this!

**Comment by dylanexpensify** [comment]
@0xmiroslav can we get that review 🙇‍♂️

**Comment by trjExpensify** [comment]
@Gonals heads up, there are conflicts to resolve here! 

@hoangzinh @0xmiroslav [on this](https://github.com/Expensify/App/pull/29790#issuecomment-1781436513) question, I think your PR can do that and apply it in this flow as well.

**Comment by Gonals** [comment]
> @Gonals just wanna confirm that this PR won't implement the confirmation modal on dismiss of the error, as @trjExpensify mentioned in his comment here [#29790 (comment)](https://github.com/Expensify/App/pull/29790#issuecomment-1771037336), will it?

Correct!

**Comment by Gonals** [comment]
Conflicts solved!

**Review by 0xmiroslav — COMMENTED** [review]
Conflicts again

File: src/components/DotIndicatorMessage.js
Comment (Gonals) [reply src/components/DotIndicatorMessage.js:94]: Ah, good call. I think I used to have more stuff in there.

**Comment by Gonals** [comment]
Comments addressed, @0xmiroslav!

**Comment by dylanexpensify** [comment]
Bump @0xmiroslav!

**Comment by trjExpensify** [comment]
👋 @0xmiroslav are you able to give us an ETA on this please?

**Comment by trjExpensify** [comment]
Thank you!

File: src/components/DotIndicatorMessage.js
Comment (Gonals) [reply src/components/DotIndicatorMessage.js:54]: Seems like something changed recently in how we display errors and this was no longer working. It should be fixed now, @0xmiroslav

File: src/libs/ErrorUtils.ts
Comment (Gonals) [reply src/libs/ErrorUtils.ts:46]: Yep

**Comment by Gonals** [comment]
@0xmiroslav, console errors fixed. The clickable thing is by design 😁

**Comment by Gonals** [comment]
> After this change, app will crash when upload receipt and fails in new version (after this PR) and then go back to old version (before this PR) because of new error type. Hope this would be fine.
> 
> Screenshot 2023-11-14 at 2 15 28 pm

Taking a look! I didn't see this right now, but I'm going to update everything to double check

**Comment by Gonals** [comment]
> After this change, app will crash when upload receipt and fails in new version (after this PR) and then go back to old version (before this PR) because of new error type. Hope this would be fine.
> 
> Screenshot 2023-11-14 at 2 15 28 pm

@0xmiroslav, can you give it another try? This doesn't seem to be crashing for me

**Comment by Gonals** [comment]
> I tried to download file on desktop but not working. Nothing happens
> 
>  Screen.Recording.2023-11-14.at.4.35.09.pm.mov

Hmmm. You are right. 
This used to work fine. Let's see if I can figure out the issue.

**Comment by Gonals** [comment]
> > I tried to download file on desktop but not working. Nothing happens
> > Screen.Recording.2023-11-14.at.4.35.09.pm.mov
> 
> Hmmm. You are right. This used to work fine. Let's see if I can figure out the issue.

Seems to be a problem it has with files in temporary memory. It has the same problem with receipts if the receipt has not been fully uploaded to our DB.

**Comment by Gonals** [comment]
@0xmiroslav. Fixed. It tried to open it in a different tab, which we don't have on Desktop.

File: src/CONST.ts
Comment (Gonals) [reply src/CONST.ts]: Sure. Updated!

**Comment by Gonals** [comment]
Done! The conflicts were someone else fixing the download issue in Desktop 😆

**Comment by dylanexpensify** [comment]
Hmm good catch - might be worth a follow up polish issue as LOW, thoughts
Alberto?

**Comment by dylanexpensify** [comment]
Yep, fair fair!

**Comment by trjExpensify** [comment]
>That iOS scroll bug is not caused by this PR. Already happening on main. That's why I suggested out of scope

Okay, really helpful if you clarify that haha. I agree that should get its own issue then. 

>https://github.com/Expensify/App/pull/29790#issuecomment-1816251869
>But this should be handled here.
>@trjExpensify @dylanexpensify any feedback?

Yes

**Review by 0xmiroslav — APPROVED** [review]

**Comment by trjExpensify** [comment]
>Jut to confirm. You mean that's expected? 😀

No, I think it should be handled here shouldn't it?

**Comment by Gonals** [comment]
> So the expected behavior is that dismissing any error should dismiss all errors related to that request, right?

Hmmm. Probably. Taking a look at other similar flows to see what happens

**Comment by Gonals** [comment]
> @Gonals error shows in both places - ReportPreview and MoneyRequestPreview The current behavior is that dismissing one doesn't dismiss another. This means that to remove this request completely, need to dismiss error twice. Is it expected?
> 
>  Screen.Recording.2023-11-17.at.12.50.33.pm.mov

This seems to be the existing behavior with any money request errors. I don't think it is what we want, but it isn't really part of this PR either. I will investigate and follow up on this to see if/how we want to fix it, but we don't need to hold the PR because of it @0xmiroslav

**Comment by Gonals** [comment]
@marcochavezf, @nkuoch, c+ approved! All yours! Let's get this done!

File: src/languages/es.ts
Comment (marcochavezf) [inline src/languages/es.ts:588]: NAB: I think it sounds a bit better
```suggestion
            loseFileMessage: 'o descarta este error y pierde el archivo',
```

**Review by marcochavezf — DISMISSED** [review]
The code LGTM 👍🏽 all yours @nkuoch

File: src/libs/actions/IOU.js
Comment (nkuoch) [inline src/libs/actions/IOU.js:347]: As these 3 lines are duplicated several times, can we create a method like `getMicroSecondIouError(receipt)`?

Comment (Gonals) [reply src/libs/actions/IOU.js:347]: Hmmm. Is it worth it? We'll still have the call 3 times (1 line instead of 2) 🤷

File: src/languages/en.ts
Comment (nkuoch) [inline src/languages/en.ts:593]: NAB - When first looking at it, I wasn't sure if "Save the file" meant "Downloading" it, or "Retrying to save it Expensify side" - Maybe renaming it it "Download the file" would remove this confusion?

File: src/libs/actions/IOU.js
Comment (nkuoch) [inline src/libs/actions/IOU.js:2986]: The idea is to pass the receipt to it, and to also use the same method for the 2 last blocks.

Comment (Gonals) [reply src/libs/actions/IOU.js]: Yep. I'm having issues testing today (onyx is doing weird stuff). I'll try again tomorrow!

**Review by nkuoch — APPROVED** [review]

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]

File: src/languages/es.ts
Comment (Gonals) [reply src/languages/es.ts:589]: I'm from Spain! I decided to leave both as they are on purpose 😁

**Review by 0xmiroslav — APPROVED** [review]

File: src/languages/es.ts
Comment (Gonals) [reply src/languages/es.ts:590]: I think using "archivo" twice in a row doesn't quite work: "Guarda el archivo o descarta este error y pierde el archivo" is a bit redundant

File: src/components/DotIndicatorMessage.js
Comment (blazejkustra) [inline src/components/DotIndicatorMessage.js:94]: @Gonals I'm working on `DotIndicatorMessage` TS migration and I while resolving conflicts I noticed ths logic change:

```ts
isReceiptError(message) ? (
                        <PressableWithoutFeedback
...
```

👇  With these lines `sortedMessages` is an array of strings, so I don't see how `isReceiptError` will ever return true in the code above, therefore the whole logic with `PressableWithoutFeedback` is unnecessary in my opinion 🤔

```ts
    const sortedMessages = _.chain(props.messages)
        .keys()
        .sortBy()
        .map((key) => props.messages[key])
        // Using uniq here since some fields are wrapped by the same OfflineWithFeedback component (e.g. WorkspaceReimburseView)
        // and can potentially pass the same error.
        .uniq()
        .map((message) => Localize.translateIfPhraseKey(message))
        .value();
```

**Comment by kavimuru** [comment]
@Gonals We are blocked as per @0xmiroslav comment. How should we proceed with this PR?

https://github.com/Expensify/App/assets/43996225/fcf230a5-bf03-4f16-9ebb-606c4eb17392

File: src/components/DotIndicatorMessage.js
Comment (Victor-Nyagudi) [inline src/components/DotIndicatorMessage.js:108]: @Gonals The error text next to red dot indicators is meant to be red to be consistent with other error messages in the app. 

With your current change here, the error text will revert back to light-greyish green. 

expensify-new-error-text-bug-introduced

I just worked on an [issue](https://github.com/Expensify/App/issues/29812) fixing this and just luckily stumbled on the regression from the video of another issue I was looking at.

**Pull request merged**
